feat: added checkout session

Niki2k1 opened the pull request:
## What kind of change does this PR introduce?

This PR adds sync for the [checkout sessions](https://docs.stripe.com/api/checkout/sessions) (and [line items](https://docs.stripe.com/api/checkout/sessions/line_items))

## Additional context

I created two new migrations (one for the checkout sessions and one for the line items.

I am not yet using this in production, so I am a little bit cautious. My tests worked fine though.

revoks commented:
This feature is something we urgently need on our side as well. 
Would love to see this merged - thanks so much for your work!

kevcodez commented on packages/sync-engine/src/stripeSync.ts:
```diff
@@ -886,6 +929,38 @@ export class StripeSync {
+    )
+
+    // Upsert line items into separate table
+    const lineItemsPromises = checkoutSessions
```
This should be a single upsert and not one upsert per line item

Are line items sent via the webhook by default? Otherwise the checkout session line items would never be there for a regular webhook

kevcodez commented:
@Niki2k1 yeah makes sense, i think a check session without persisting the line items themselves would not add too much value.

I think it would be best to call `stripe.checkout.sessions.listLineItems` (and set a high limit of 100) if no line items is provided - we also have an expansion feature that can be used (see `expandEntity`).

So
- If line items is already present, use `expandEntity`
- If line items are not present, pull from API using `listLineItems`

jackkru69 commented:
@Niki2k1 Hi, are you going to refine this PR?

Niki2k1 commented:
@jackkru69 yes, I will implement the changes and update the pr, but at earliest next week. Can't promise anything though.

Niki2k1 commented:
I merged main and resolved some conflicts and also added the new lastSyncedAt logic

jackkru69 commented:
@kevcodez Could you check it, pls?

andrey-utkin commented:
Hi @Niki2k1 , thanks for your very useful contribution!

I have just tried a container image built from your feature branch, with database containing products of backfill by mainline codebase. I also retried from scratch. I use backfill with `POST /sync`. I got this error. Maybe the reason is that we might encounter more than one event about the same session, and see the same (by id) line item more than once?

```
stripe-sync-1  | {"level":"info","time":"2025-08-29T22:12:15.976Z","pid":1,"hostname":"e92d0932b284","msg":"Syncing checkout sessions"}
stripe-sync-1  | {"level":"info","time":"2025-08-29T22:12:15.976Z","pid":1,"hostname":"e92d0932b284","msg":"Fetching items to sync from Stripe"}
stripe-sync-1  | {"level":"info","time":"2025-08-29T22:12:58.862Z","pid":1,"hostname":"e92d0932b284","msg":"Upserting 1819 items"}
postgres-1     | 2025-08-29 22:12:59.458 UTC [54] ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification
postgres-1     | 2025-08-29 22:12:59.458 UTC [54] STATEMENT:  INSERT INTO "stripe"."checkout_session_line_items" (
postgres-1     |                "id","object","adjustable_quantity","amount_subtotal","amount_total","currency","description","discounts","price","quantity","taxes","checkout_session", "last_synced_at"
postgres-1     |              )
postgres-1     |              VALUES (
postgres-1     |                $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12, $13
postgres-1     |              )
postgres-1     |              ON CONFLICT (id) DO UPDATE SET
postgres-1     |                "id" = EXCLUDED."id","object" = EXCLUDED."object","adjustable_quantity" = EXCLUDED."adjustable_quantity","amount_subtotal" = EXCLUDED."amount_subtotal","amount_total" = EXCLUDED."amount_total","currency" = EXCLUDED."currency","description" = EXCLUDED."description","discounts" = EXCLUDED."discounts","price" = EXCLUDED."price","quantity" = EXCLUDED."quantity","taxes" = EXCLUDED."taxes","checkout_session" = EXCLUDED."checkout_session",
postgres-1     |                last_synced_at = $14
postgres-1     |              WHERE "checkout_session_line_items"."last_synced_at" IS NULL
postgres-1     |                 OR "checkout_session_line_items"."last_synced_at" < $15;

```

andrey-utkin commented:
I guess `ON CONFLICT (id)` would work if `checkout_session_line_items` `id` itself was a primary key.

I am not even sure what this primary key specification means: `primary key ("checkout_session", "id")`

Indexes/keys end up this way:
```
Indexes:
    "checkout_session_line_items_pkey" PRIMARY KEY, btree (checkout_session, id)
    "stripe_checkout_session_line_items_price_idx" btree (price)
    "stripe_checkout_session_line_items_session_idx" btree (checkout_session)
Foreign-key constraints:
    "checkout_session_line_items_checkout_session_fkey" FOREIGN KEY (checkout_session) REFERENCES checkout_sessions(id)
```

Niki2k1 commented:
I added a test (was a good idea because I found multiple issues like missing fields)
For now, it's only an insert. Do we need other cases as well?

I also added `on delete cascade` on the lineItems reference fields.
So when a checkout session is deleted, the line items are getting deleted as well. For the prices reference, that's not really needed, as prices and checkout_sessions can't be deleted in stripe. But was useful in testing (for cleanup) and I don't see what would speak against it.

kevcodez reviewed (commented):
Overall LGTM with the tests, two small things I noticed that would have to be addressed so we can merge this 👍

Niki2k1 commented on packages/sync-engine/src/stripeSync.ts:
```diff
@@ -550,6 +572,10 @@ export class StripeSync {
       return this.stripe.refunds.retrieve(stripeId).then((it) => this.upsertRefunds([it]))
+    } else if (stripeId.startsWith('cs_')) {
+      return this.stripe.checkout.sessions
+        .retrieve(stripeId)
```
The line_items are getting filled in the upsert itself via their own endpoint. They are not available in the [/v1/checkout/sessions/:id](https://docs.stripe.com/api/checkout/sessions/retrieve) endpoint.
correct me if I am wrong

kevcodez replied:
Aren't you doing the expansion in here too? https://github.com/supabase/stripe-sync-engine/pull/172/files#diff-95fc8d6564ba293c6ce4a37ea33f73126fab928a0311b74c1a56b5b2d19d0bdeR122

Niki2k1 replied:
I will remove it there then as well 😅
was left over from when I started.
I moved over to the `fillCheckoutSessionsLineItems` which uses the [/v1/checkout/sessions/:id/line_items](https://docs.stripe.com/api/checkout/sessions/line_items) endpoint

Niki2k1 commented:
* removed line_items expands
* backfill prices (needed for the new price table reference) to ensure prices are there before insertion of lineItems
* using auto pagination feature
* fix tests

also tested again with my live data

kevcodez commented:
Nice! Thanks for the contribution - will trigger a release now
